File: bst.py
class Node:
    def __init__(self, val):
        self.val = val 
        self.left = self.right = None
    
class BSTree :
    def __init__(self):
        self.root = None 

    def __str__(self):
        return "my tree"

    def BFT(self):
        queue = []
        queue.append(self.root)
        while len(queue) != 0 :
            cur = queue[0]
            queue = queue[1:]

            print(cur.val, "->")
            if cur.left != None:
                queue.append(cur.left)
            if cur.right != None:
                queue.append(cur.right)

    def add(self, val):
        n = Node(val)

        if self.root == None :
            self.root = n
            return 
        
        cur = self.root
        while cur != None :
            if cur.val < val :
                if cur.right == None :
                    cur.right = n 
                    break 
                cur = cur.right
            else:
                if cur.left == None :
                    cur.left = n
                    break 
                cur = cur.left

    
t = BSTree()
t.add(1)
t.add(2)
t.add(3)
t.add(4)
t.BFT()

File: bst.kt

fun main(args: Array<String>) {
    println("Hello")

    var t = Tree<Int>()

    t.add(1)
    t.add(2)
    t.add(3)
    t.add(4)
    t.add(5)

    t.BFT()

    var t2 = Tree<Int>()
    t2.add(10)
    t2.add(11)
    t2.add(12)
    t2.add(1)
    t2.add(2)
    t2.BFT()
}

class Node<T: Comparable<T>>(val value: T?=null) {
    var left: Node<T>? = null
    var right: Node<T>? = null
}

class Tree<T : Comparable<T>>() {
    var root: Node<T>? = null

    fun add(item: T): Unit {
        var n: Node<T> = Node<T>(item)

        if (root == null) {
            root = n
            return
        }

        var cur = root

        while (cur != null) {
            if (item > cur.value!!) {
                if (cur.right == null) {
                    cur.right = n
                    break
                }
                cur = cur.right
            } else {
                if (cur.left == null) {
                    cur.left = n
                    break
                }
                cur = cur.left
            }
        }
    }

    fun BFT() {
        if (root == null) return
        var q = mutableListOf<Node<T>>(root!!)

        while(q.size!=0){
            var cur = q.removeAt(0)
            println("${cur.value} ->")

            if (cur.left != null) {
                q.add(cur.left!!)
            }
            if (cur.right != null) {
                q.add(cur.right!!)
            }
        }
    }
}
